qmi/instruments/tenma/psu_72.py

```diff
@@ -0,0 +1,490 @@
+            )
+        elif transport.startswith("udp"):
+            self._transport = create_transport(transport)
+        else:
+            raise QMI_InstrumentException(f"Transport type {transport} not valid!")
+
+    def _send(self, cmd: str) -> None:
+        """Writes a command to the Tenma, just an ascii string
+
+        Parameters:
+            cmd:    The command to write to the device.
+        """
+        _logger.info("[%s] Sending message %s", self._name, cmd)
+        self._transport.write(bytes(cmd, "ascii"))
+        sleep(.02)  # Sleep some bits to let the message arrive
+
+    def _read(self, timeout: float = 0.2) -> str:
+        """Read the response from the device, try until response received or time-out.
+
+        Parameters:
+            timeout: The time-out value for reading. Default is 200ms.
+
+        Returns:
+            response: Decoded response string.
+        """
+        response = self._transport.read_until_timeout(self.BUFFER_SIZE, timeout)
+        _logger.info("[%s] Read message %s", self._name, response)
+        return response.decode('ascii')
+
+    def _enable_output(self, output: bool, command: str) -> None:
+        """Internal function to further communicate the output state to instrument.
+
+        Parameters:
+            output: The expected output state.
+            command: The enable command string.
+        """
+        self._send(command)
+        status = self.get_status()
+        if status["OutputEnabled"] is not output:
+            _logger.exception(
+                "[%s] Tried to set output with %s but got %s", self._name, command, status["OutputEnabled"]
+            )
+            raise QMI_InstrumentException("Setting output failed for an unknown reason")
+
+    @rpc_method
+    def open(self) -> None:
+        self._check_is_closed()
+        _logger.info("Opening connection to %s", self._name)
+        self._transport.open()
+        super().open()
+
+    @rpc_method
+    def close(self) -> None:
+        self._check_is_open()
+        _logger.info("Closing connection to %s", self._name)
+        super().close()
+        self._transport.close()
+
+    @rpc_method
+    def get_idn(self) -> QMI_InstrumentIdentification:
+        """Get instrument identification info.
+
+        Returns:
+            QMI_InstrumentIdentification: NamedTuple with vendor, model, serial number and SW version info.
+        """
+        self._send("*IDN?")
+        idn = self._read()
+        # Find all relevant info from response. Should be of format TENMA 72-2535 SN:1231345 V2.0
+        pattern = re.compile(r'(\b\w+\b[^ ]*|\b72[^ ]*|V[^ ]*|SN:[^ ]*)')
+
+        # Find all matches in the input string
+        matches = pattern.findall(idn)
+
+        # Extract the relevant parts from the matches
+        vendor = matches[0] if matches else None
+        model = next((match for match in matches[1:] if match.startswith('72')), None)
```

Comment:
would be nice to have a test for the default value of None for these variables

Reply:
Hmm... For `vendor`, `model`, `version` nothing (exciting) happens, but indeed `serial` would cause an `AttributeError` in the `serial=serial.strip()[3:],`. I think a check on this is needed to return `None` instead of excepting on it. Then just returning `None` for all `QMI_InstrumentIdentification` values is a reasonable test.

Reply:
... unless it would be preferable to raise an exception in that case?

Reply:
I'm happy with returning None, but this scenario is not tested yet

Reply:
Well I made it to raise an exception, if that's OK.